framework/src/main/java/org/checkerframework/framework/type/AbstractViewpointAdapter.java

```diff
@@ -0,0 +1,492 @@
+    public void viewpointAdaptTypeParameterBounds(
+            AnnotatedTypeMirror receiverType,
+            List<AnnotatedTypeParameterBounds> typeParameterBounds) {
+
+        List<AnnotatedTypeParameterBounds> adaptedTypeParameterBounds = new ArrayList<>();
+        for (AnnotatedTypeParameterBounds typeParameterBound : typeParameterBounds) {
+            AnnotatedTypeMirror adaptedUpper =
+                    combineTypeWithType(receiverType, typeParameterBound.getUpperBound());
+            AnnotatedTypeMirror adaptedLower =
+                    combineTypeWithType(receiverType, typeParameterBound.getLowerBound());
+            adaptedTypeParameterBounds.add(
+                    new AnnotatedTypeParameterBounds(adaptedUpper, adaptedLower));
+        }
+
+        typeParameterBounds.clear();
+        typeParameterBounds.addAll(adaptedTypeParameterBounds);
+    }
+
+    /**
+     * Viewpoint adapt declared type to receiver type, and return the result atm
+     *
+     * @param receiver receiver type
+     * @param declared declared type
+     * @return {@link AnnotatedTypeMirror} after viewpoint adaptation
+     */
+    protected AnnotatedTypeMirror combineTypeWithType(
+            AnnotatedTypeMirror receiver, AnnotatedTypeMirror declared) {
+        assert receiver != null && declared != null;
+
+        AnnotatedTypeMirror result = declared;
+
+        if (receiver.getKind() == TypeKind.TYPEVAR) {
+            receiver = ((AnnotatedTypeVariable) receiver).getUpperBound();
+        }
+        AnnotationMirror receiverAnnotation = extractAnnotationMirror(receiver);
+        if (receiverAnnotation != null) {
+            result = combineAnnotationWithType(receiverAnnotation, declared);
+            result = substituteTVars(receiver, result);
+        }
+
+        return result;
+    }
+
+    /**
+     * Extract modifier from {@link AnnotatedTypeMirror}.
+     *
+     * @param atm AnnotatedTypeMirror from which modifier is extracted
+     * @return modifier extracted
+     */
+    protected abstract AnnotationMirror extractAnnotationMirror(AnnotatedTypeMirror atm);
+
+    /**
+     * Sub-procedure to combine receiver modifiers with declared types. Modifiers are extracted from
+     * declared types to furthur perform viewpoint adaptation only between two modifiers.
+     *
+     * @param receiverAnnotation receiver modifier
+     * @param declared declared type
+     * @return {@link AnnotatedTypeMirror} after viewpoint adaptation
+     */
+    protected AnnotatedTypeMirror combineAnnotationWithType(
+            AnnotationMirror receiverAnnotation, AnnotatedTypeMirror declared) {
+        if (declared.getKind().isPrimitive()) {
+            AnnotatedPrimitiveType apt = (AnnotatedPrimitiveType) declared.shallowCopy();
+
+            AnnotationMirror declaredAnnotation = extractAnnotationMirror(apt);
+            AnnotationMirror resultAnnotation =
+                    combineAnnotationWithAnnotation(receiverAnnotation, declaredAnnotation);
```

Comment:
seems like declaredAnnotation is only ever used in the argument of combineAnnotationWithAnnotation, can eliminate this variable all together and nest the call to extractAnnotationMirror

Reply:
Done